src/printer/layers.js

```diff
@@ -111,21 +109,58 @@ function createLayerXYZ(layerSpec, rootFrameState) {
       return frameState.tileQueue.getTilesLoading();
     }, true),
     map(() => {
-      const progress =
-        1 -
-        Object.keys(frameState.tileQueue.queuedElements_).length / tileCount;
+      let loadedTilesCount = Object.keys(frameState.tileQueue.queuedElements_)
+        .length;
+
+      let progress = 1 - loadedTilesCount / tileCount;
+
+      // this is to make sure all tiles have finished loading before completing layer
+      if (progress === 1 && frameState.tileQueue.getTilesLoading() > 0) {
```

Comment:
So to know the advancement of a print, we look at
1. how many tiles are queued
2. how many tiles are currently loading

Initially OL will queue as many tiles as it needs for rendering the whole extent, so let's say 100. There will be no tiles currently loading. `tileCount` will be set at 100.

Then every 500ms we reprioritize the tile queue and ask for more tiles to load. The amount of queued tiles will go down for each tiles being loaded: 99, 98, 97...

Then at some point there will be no tiles left in the queue, BUT there still may be tiles loading and not ready yet. Because the progress is computed only with the amount of tiles queued, it would be 1 in that case. To avoid the layer being printed and sent back to the main thread we need to wait for the count of loading tiles to reach zero, hence this condition: if there are still tiles loading and progress is one, set progress to 0.999. Tiles usually load very quickly anyway so this is invisible to the user.